src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1609,6 +1676,140 @@ RebalanceTableShards(RebalanceOptions *options, Oid shardReplicationModeOid)
 }
 
 
+static bool
+HasConcurrentRebalance(void)
+{
+	LOCKTAG locktag = { 0 };
+	SET_LOCKTAG_CITUS_OPERATION(locktag, CITUS_REBALANCE);
+	LockAcquireResult locked = LockAcquire(&locktag, ExclusiveLock, false, true);
+	if (locked == LOCKACQUIRE_NOT_AVAIL)
+	{
+		return true;
+	}
+	LockRelease(&locktag, ExclusiveLock, false);
+
+
+	if (HasNonTerminalJobOfType("rebalance", NULL))
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+static void
+ErrorOnConcurrentRebalance(void)
+{
+	if (HasConcurrentRebalance())
+	{
+		/* TODO find hint/detail messages to show */
+		ereport(ERROR, (errmsg("a rebalance is already running")));
+	}
+}
+
+
+/*
+ * RebalanceTableShardsBackground rebalances the shards for the relations
+ * inside the relationIdList across the different workers. It does so using our
+ * background job+task infrastructure.
+ */
+static void
+RebalanceTableShardsBackground(RebalanceOptions *options, Oid shardReplicationModeOid)
+{
+	ErrorOnConcurrentRebalance();
+
+	if (list_length(options->relationIdList) == 0)
+	{
+		return;
+	}
+
+	Oid relationId = InvalidOid;
+	char *operationName = "rebalance";
+	if (options->drainOnly)
+	{
+		operationName = "move";
+	}
+
+	foreach_oid(relationId, options->relationIdList)
+	{
+		AcquireRebalanceColocationLock(relationId, operationName);
+	}
+
+	List *placementUpdateList = GetRebalanceSteps(options);
+
+	if (list_length(placementUpdateList) == 0)
+	{
+		return;
+	}
+
+	/* find the name of the shard transfer mode to interpolate in the scheduled command */
+	Datum shardTranferModeLabelDatum =
+		DirectFunctionCall1(enum_out, shardReplicationModeOid);
+	char *shardTranferModeLabel = DatumGetCString(shardTranferModeLabelDatum);
+
+	if (list_length(placementUpdateList) == 0)
+	{
+		return;
+	}
+
+	/* schedule planned moves */
+	int64 jobId = CreateBackgroundJob("rebalance", "Rebalance colocation group ...");
+
+	/* buffer used to construct the sql command for the tasks */
+	StringInfoData buf = { 0 };
+	initStringInfo(&buf);
+
+	/*
+	 * Currently we only have two tasks that any move can depend on:
+	 *  - replicating reference tables
+	 *  - the previous move
+	 *
+	 * prevJobIdx tells what slot to write the id of the task into. We only use both slots
+	 * if we are actually replicating reference tables.
+	 */
+	int64 prevJobId[2] = { 0 };
+	int prevJobIdx = 0;
+
+	if (true /* TODO check if replicating reference tables is required */)
+	{
+		/*
+		 * Reference tables need to be copied to (newly-added) nodes, this needs to be the
+		 * first task before we can move any other table.
+		 */
+		appendStringInfo(&buf,
+						 "SELECT pg_catalog.replicate_reference_tables(%s)",
```

Comment:
It would be a little nicer to schedule in the individual copy commands given that we'll go maintenance daemon spawns monitor -> monitor spawns executer -> executer runs replicate_reference_tables which connects to localhost -> localhost connection run citus_copy_shard_placement -> citus_copy_shard_placement probably connects to localhost to set up replication from coordinator. The inception runs pretty deep.

Not sure how hard it is to get the individual citus_copy_shard_placement commands. Perhaps easier now that citus_copy_shard_placement is idempotent?).

Reply:
If copy is idempotent it will probably work. There could be some time in between a rebalance triggered and all copies executed, due to a concurrent `EnsureReferenceTablesExistOnAllNodes`.

For the copy commands we will probably need to reuse some of the reference table replication logic to find the nodes where the reference tables are missing.

Copy does require a source node as well iirc. Will probably use the same heuristics as this function, but that still would allow -albeit a very edgecase- for the reference tables to be removed from the source before the copy runs. But I still think that running copies here is a nicer thing compared to the backend dance you describe above.

It will also implicitly add the check if the tables still need to be copied :D

Reply:
Kind of seems like a bug. Not much preventing users from calling copy_shard_placement on reference tables themselves I think.

Reply:
Jelte is correct here. Manually copying reference table groups has a chance that concurrently reference tables are being created/dropped when _not_ done via `replicate_reference_tables`. This was already existing before we allowed concurrent creates/drops, and is the same as #6066 , except is has only been solved for hash distributed tables here.

To solve this correctly we need to overhaul how reference tables are created when they colocate with existing reference tables so that we can remove ensure reference tables exists from most codepaths where correct locking would cause undetectable deadlocks.